evolver/device.py

```diff
@@ -70,26 +74,49 @@ def schema(self):
             "controllers": [{"kind": str(type(a)), "config": a.Config.model_json_schema()} for a in self.controllers],
         }
 
+    def _loop_exception_wrapper(self, callable, message="unknown") -> bool:
+        try:
+            callable()
+            return False
+        except Exception:
+            self.logger.exception(f"Error in loop: {message}:")
+            if self.raise_loop_exceptions:
+                raise
+            return True
+
     def read_state(self):
+        read_error = False
         for name, device in self.sensors.items():
-            device.read()
+            err = self._loop_exception_wrapper(device.read, f"reading device {name}")
+            read_error = read_error or err
             self.last_read[name] = time.time()
             self.history.put(name, device.get())
+        return read_error
 
     def evaluate_controllers(self):
-        for controller in self.controllers:
-            controller.run()
+        errs = [self._loop_exception_wrapper(c.run, f"updating controller {c}") for c in self.controllers]
+        return any(errs)
 
     def commit_proposals(self):
-        for device in self.effectors.values():
-            device.commit()
+        errs = [
+            self._loop_exception_wrapper(d.commit, f"committing proposals for {d}") for d in self.effectors.values()
+        ]
+        return any(errs)
 
     def loop_once(self):
-        self.read_state()
-        # for any hardware awaiting calibration, call calibration update method here
+        read_error = self.read_state()
+        if read_error and self.skip_control_on_read_failure:
+            self.logger.info("Skipping control loop due to read error")
+            return
         if self.enable_control:
-            self.evaluate_controllers()
-            self.commit_proposals()
+            control_error = self.evaluate_controllers()
+            if control_error and self.abort_on_control_errors:
+                self.abort()
+                return
```

Comment:
@amitschang Sorry, I should have noted this before - what's the rationale for not wanting to actually ``raise`` here? I would have expected this to raise the error not just ``return``?

Reply:
Since we captured the errors, logged and aborted, we can assume we are in a "clean" state (from app perspective). If the user were to re-enable the flags cause they deem it OK, then the application could continue.

We can raise here, but then would need to capture it in the webapp loop activity, but then I think we might need to differentiate between errors raised by `_raise_loop_exceptions` and others.

Reply:
Sure, but the web app is its own responsibility. However, when using this programmatically, we would expect it to raise, no? I feel like the alternative is wiping python's exception handling functionality clean off the map.

Reply:
OK. I guess we should then raise an exception if any captured error happened within the loop, not just if the thing was aborted.

Reply:
I might be wrong, but I think that only hangs on the semantics of "abort" being name-mangled into the flags, right? Otherwise, I'm not following your point and feel like I missing something for that to be true.

Ignoring the "abort" name-mangle for a mo, I see the flags as more of a "continue-don't-raise" (inverted bool on what we have here). So to me, the ``raise`` is _the_ point here, however, we gracefully "abort" because an exception was raised so that hardware state can be "reset".  The "abort" isn't the focus from my perspective.

Again, as noted before, this screams for just using context-managers to handle the abort and thus keeping the focus of these flags and conditional checks on the option of whether or not to raise etc.

Reply:
what do you mean by "name-mangled"? 

I might have misinterpreted what you said before then. Do you mean that because we choose to abort on the erorr, it should raise some exception, not that it should raise to indicate any error was encounted. How would it be in the case of skip-on-read? Or strictly for abort since it is like a termination condition?

Reply:
something like latest change?

Reply:
RE name-mangling: ``abort_on_control_errors`` & ``abort_on_commit_errors``.

>How would it be in the case of skip-on-read?

Skip-on-read is ok, because we're explicitly skipping and continuing.

>Or strictly for abort since it is like a termination condition?

Exactly, the ``abort`` terminates (the app can deal with its own continuation semantics/functionality) however, these flags are in the ``Evolver.Config`` and should be agnostic to the app.

My main concern here is that if we abort and just return and continue at this layer, for these three core functions, aren't we nuking all exceptions from being raised (unless ``raise_loop_exceptions==True``). I think this is the point that we should focus on, if I'm wrong then we should be ok. But it seems to me that this PR does away with exceptions being raised, since the default for ``raise_loop_exceptions`` is ``False`` with core focus being shifted to the use of the other three flags which absorb raised exceptions.

Perhaps, my misunderstanding is with the following, I see these flags as completely overlapping and not different functionality/semantics, but _just_ granularity. I.e., where this def is true: raise_loop_exceptions := abort_on_control_errors & abort_on_commit_errors & ~skip_control_on_read_failure.